crates/autopilot/src/run_loop.rs

```diff
@@ -121,9 +124,68 @@ impl RunLoop {
         driver: &Driver,
         _solution: &solve::Response,
     ) -> Result<()> {
-        let request = execute::Request { auction_id: id };
+        let request = execute::Request {
+            auction_id: id,
+            tag: id.to_be_bytes().into(),
+        };
         let _response = driver.execute(&request).await.context("execute")?;
-        // TODO: Wait for transaction to be mined or deadline to be reached.
+        // TODO: React to deadline expiring.
+        self.wait_for_settlement_transaction(&request.tag)
+            .await
+            .context("wait for settlement transaction")?;
         Ok(())
     }
+
+    /// Tries to find a `settle` contract call with calldata ending in `tag`.
+    ///
+    /// Returns None if no transaction was found within the deadline.
+    pub async fn wait_for_settlement_transaction(&self, tag: &[u8]) -> Result<Option<Transaction>> {
+        // Start earlier than current block because there might be a delay when receiving the
+        // Solver's /execute response during which it already started broadcasting the tx.
+        let start_offset = 10;
+        let max_wait_time = 20;
+        let current = self.current_block.borrow().number;
+        let start = current.saturating_sub(start_offset);
+        let deadline = current.saturating_add(max_wait_time);
+        tracing::debug!(%current, %start, %deadline, ?tag, "waiting for tag");
+
+        // Use the existing event indexing infrastructure to find the transaction. We query all
+        // settlement events in the block range to get tx hashes and query the node for the full
+        // calldata.
+        //
+        // If the block range was large, we would make the query more efficient by moving the
+        // starting block up while taking reorgs into account. With the current range of 30 blocks
+        // this isn't necessary.
+        //
+        // We do keep track of hashes we have already seen to reduce load from the node.
+
+        let mut seen_transactions: HashSet<H256> = Default::default();
+        while self.current_block.borrow().number <= deadline {
```

Comment:
This drops the `burrow()` right? According to the docs:

> Outstanding borrows hold a read lock. This means that long lived borrows could cause the send half to block. It is recommended to keep the borrow as short lived as possible.

Maybe it might make sense to:
```suggestion
        while {
            self.current_block.borrow().number <= deadline
        } {
```

Or more explicitly:
```suggestion
        while {
            let current = self.current_block.borrow();
            let within_deadline = current.number <= deadline
            drop(current);
            within_deadline
        } {
```

(The reason being I remember something about this with `match` statements and don't know the exact semantics here).

Reply:
I ran into that [problem](https://github.com/cowprotocol/services/pull/365) before.
You could also turn the block stream into an actual stream with [`into_stream()`](https://github.com/cowprotocol/services/blob/main/crates/shared/src/current_block.rs#L119-#L121). Then you wouldn't have to manage the timeout yourself.

Reply:
Temporaries in the while loop condition don't get their lifetime extended for the while itself so it isn't a problem here. I think I'll still change it to be more obviously correct.

Reply:
Cool! Just wanted to be sure.

> Temporaries in the while loop condition

I guess this only applies to `while boolean_condition` and not `while let`?

Reply:
I *think* it is about whether you can refer to the temporary later. With `match` and `while let` you are binding something that you can then make use of. For these cases the temporary stays alive because there could be lifetime parameters from the temporary into the binding. With `while` you can't do this so there is no reason for the temporary to stay alive. `while` can be transformed into `loop`, where it also makes sense that the temporary wouldn't stay alive.